src/resources/ble_esp32.cc

```diff
@@ -763,307 +1724,326 @@ PRIMITIVE(advertise_config) {
   return process->program()->null_object();
 }
 
 PRIMITIVE(advertise_stop) {
-  ARGS(BLEResourceGroup, group);
-
-  if (BLEResourceGroup::is_advertising()) {
+  ARGS(BLEPeripheralManagerResource, peripheral_manager);
+  if (BLEPeripheralManagerResource::is_advertising()) {
     int err = ble_gap_adv_stop();
-    if (err != ESP_OK) {
-      return Primitive::os_error(err, process);
+    if (err != BLE_ERR_SUCCESS) {
+      return nimle_stack_error(process,err);
     }
   }
+  peripheral_manager->set_advertising_started(false);
 
   return process->program()->null_object();
 }
 
-PRIMITIVE(connect) {
-  ARGS(BLEResourceGroup, group, Blob, address, GATTResource, gatt);
-
-  uint8_t own_addr_type;
-
-  int err = ble_hs_id_infer_auto(0, &own_addr_type);
-  if (err != ESP_OK) {
-    return Primitive::os_error(err, process);
-  }
+PRIMITIVE(add_service) {
+  ARGS(BLEPeripheralManagerResource, peripheral_manager, Blob, uuid)
 
-  ble_addr_t addr{};
-  addr.type = address.address()[0];
-  memcpy_reverse(addr.val, address.address() + 1, 6);
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (proxy == null) ALLOCATION_FAILED;
+  ble_uuid_any_t ble_uuid = uuid_from_blob(uuid);
 
-  err = ble_gap_connect(own_addr_type, &addr, 3000, NULL,
-                        BLEEventSource::on_gap, gatt);
-  if (err != ESP_OK) {
-    return Primitive::os_error(err, process);
-  }
+  BLEServiceResource* service_resource =
+      peripheral_manager->get_or_create_service_resource(ble_uuid, 0,0, true);
+  if (!service_resource) MALLOC_FAILED;
+  if (service_resource->deployed()) INVALID_ARGUMENT;
 
-  return process->program()->null_object();
+  proxy->set_external_address(service_resource);
+  return proxy;
 }
 
-PRIMITIVE(get_gatt) {
-  ARGS(BLEResourceGroup, group);
+PRIMITIVE(add_characteristic) {
+  ARGS(BLEServiceResource, service_resource, Blob, raw_uuid, int, properties, int, permissions, Object, value)
 
-  ByteArray* proxy = process->object_heap()->allocate_proxy();
-  if (!proxy) ALLOCATION_FAILED;
+  if (!service_resource->peripheral_manager()) INVALID_ARGUMENT;
 
-  GATTResource* gatt = _new GATTResource(group);
-  if (!gatt) MALLOC_FAILED;
+  ByteArray* proxy = process->object_heap()->allocate_proxy();
+  if (proxy == null) ALLOCATION_FAILED;
 
-  group->register_resource(gatt);
-  proxy->set_external_address(gatt);
+  if (service_resource->deployed()) INVALID_ARGUMENT;
 
-  return proxy;
-}
+  ble_uuid_any_t ble_uuid = uuid_from_blob(raw_uuid);
 
-PRIMITIVE(request_result) {
-  ARGS(GATTResource, gatt);
+  os_mbuf* om = null;
+  Object* error = object_to_mbuf(process, value, &om);
+  if (error) return error;
 
-  if (gatt->error() == BLE_HS_ENOENT) {
-    OUT_OF_RANGE;
-  } else if (gatt->error() != 0) {
+  uint32 flags = properties & 0x7F;
+  if (permissions & 0x1) flags |= BLE_GATT_CHR_F_READ;
+  if (permissions & 0x2 && !(properties & (BLE_GATT_CHR_PROP_WRITE | BLE_GATT_CHR_F_WRITE_NO_RSP)))
```

Comment:
If the properties isn't a write, then it's an invalid argument.
But don't we need to set the flags, when the properties are write?

Also: don't we need to do the same properties test for the `write_enc` below?

Reply:
The reason this is here in the first place is due to the way that macos defines this. It could be argued that the mac implementation should adhere to the nimble, but in this case I turned it around. The code can not know from permissions&0x2 if it is a write with or without response and therefore it does not try to be clever.